Решение задания "Проект 1 - ДНК"

slebedev opened the pull request:
Надеюсь, первый блин не будет комом. =)

slebedev committed:
Решённое задание "Проект 1 - ДНК"

Надеюсь, первый блин не будет комом. =)

hmmvot commented on DNAProject/Cell.m:
```diff
@@ -0,0 +1,51 @@
+}
+
+-(NSString *)description //этот метод позволит выводить ДНК в виде NSLog(@"%@",myCell)
+{
```
return [DNA componentsJoinedByString: @""];
Так заметно проще и оптимальнее.)

slebedev replied:
Спасибо, не дошёл до такого метода, это заметно всё упрощает, да.

hmmvot commented on DNAProject/Cell.m:
```diff
@@ -0,0 +1,51 @@
+    NSString *compareString = [[NSString alloc] initWithString:[someCell description]];
+    for (int i=0; i<capacityOfDNA; i++) {
+        if ([sourceString characterAtIndex:i]!=[compareString characterAtIndex:i]) ham++;
+    }
```
Создайте @property для переменной DNA. Таким образом вы избавитель от необходимости конвертировать Cell в строку и сможете сравнивать элементы DNA для каждой клетки на прямую.

slebedev replied:
А в задании не сказано, что мы сможем менять последовательность напрямую. Всё же, идеология объектов такова, что нужно стараться делать внутренние переменные закрытыми (так меня учили, и так всегда и стараюсь делать).
Разумеется, вышесказанное применимо далеко не всегда. но в данном случае, когда у нас есть объект с последовательностью ДНК, я считаю, что не нужно давать возможность напрямую менять последовательность. Максимум - можно сделать @property (readonly) для данной переменной, но опять же, не вижу в этом практического смысла исходя из задания.

hmmvot replied:
1) Практический смысл в том, чтобы избавиться от не нужной здесь операции преобразования массива в строку.
2) @property (nonatomic, readonly) NSArray *DNA; благодаря такому property вы сможете обращаться к DNA из вне, но не сможете изменить указатель, хранящейся в DNA, и не сможете изменить массив, хранящийся в памяти по этому указателю. Вот вам поддержка идеологии объекта.

slebedev replied:
Да, я именно про эту возможность сделать свойство доступным только на чтение.

slebedev replied:
Всерьёз подумал над вашим предложением, только если переделать переменную класса в свойство (мало того, что readonly, так ещё и NSArray), то не смогу его менять в своей категории. А по заданию, именно метод категории должен модифицировать последовательность.

hmmvot replied:
Никаких проблем с категорией не будет. @property создаст геттер и сеттер, что никак не повлияет на обращение к переменным в определении класса.

slebedev replied:
readonly-свойство создаст только геттер.

aspcartman replied:
Поле Description используется исключительно для логирования. Нельзя (ну как, никто не запретит, но это моветон) его использовать для других целей
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSObject_Class/Reference/Reference.html#//apple_ref/occ/clm/NSObject/description

Проперти создать нужно, почти что по заданию - такова была тема лекций. А делать ли его readonly или нет - исключительно ваш выбор :)

slebedev replied:
Во всяком случае, вывод на экран - это по сути и есть логирование, в каком-то смысле. А метод этот был подсмотрен в книге Аарона Хиллегаса, и рекомендовался он именно для того, чтобы объект класса можно было выводить на экран. 

Я согласен, что этот материал был в лекции, но так же и сказал, почему в данном случае не захотел использовать их - лучше уж без свойства. А вообще, если даже это влияет на оценку, хотелось бы услышать комментарий @freetonik по этому поводу.

hmmvot replied:
Ваш подход замедлен тем, что вместо прямого обращения к массиву вы создаете два строковых объекта из массивов (что медленно).
И чем вам не нравится property - не понятно, т.к. property является инструментом инкапсуляции и идеологию объекта как раз поддерживает полностью.
Если же вы действительно уверенны, что в данном случае property - это зло, вы можете сделать так:

``` objective-c
if([DNA objectAtIndex: i] != [someCell->DNA objectAtIndex: i]) { ...
```

aspcartman replied:
Так вы же используете в этом месте не как логирование, а как получение даных DNA для обработки. Это очень плохо, и повлияет на оценку,  так делать нельзя.

ЗЫ: вы акуратней отправляйте с почты, посмотрите, вы цитатой заполнили всю страницу :)

slebedev replied:
Простите, впредь буду следить, чтоб при ответе не цитировалось предыдущее.

slebedev replied:
Хм, а разве так получится? Ведь переменная по умолчанию приватная... Но попробовать стоит, спасибо. 

Уже думаю написать модификатор по индексу и сделать проперти только для чтения. Надо было сразу так и делать, но подумалось, что написанное мной будет проще.

aspcartman replied:
Прошу обратить внимание, что в Obj-C все - public, и все методы - override. Т.е. методологии других языков в плане инкапсуляции и механизмов наследования к Obj-C применимы лишь через некоторую призму. Особенно на это обратить нужно внимание, если раньше вы писали на С++ или Java.

slebedev replied:
Полезное замечание, спасибо!

slebedev replied:
Кстати, нашёл "Object-C way", вернее, аналогичный способ доступа к внутреннему состоянию объекта в Objective-C. Вместо [someCell->DNA objectAtIndex:i] можно замутить [[someCell valueForKey:@"DNA"] objectAtIndex:i], пускай строка длиннее, но всю идеологию доступа к внутреннему состоянию объектов выдаёт.

hmmvot replied:
Ага, только в первом случае - прямой доступ, а во втором - посылка сообщения (если это работает так, как выглядит). И прямой доступ явно быстрее.

slebedev replied:
Фишка в том, что прямой доступ, опять же, будет работать только в пределах файла с имплементацией класса (в моём случае). Смотрите какая штука: у меня категория объявлена в файле main.m (как в примере). Если делать @property (nonatomic, readonly) NSMutableArray *DNA, то в этой категории я не смогу изменить значение самого массива _DNA, так как категория не видит такой переменной, а свойство объявлено как readonly. Тем не менее, если посылать сообщение объекту self, то всё пройдёт замечательно!

[self->_DNA replaceObjectAtIndex:generated withObject:modifiedGene]; // Error: 'Cell' does not have a member named '_DNA'

[[self valueForKey:@"_DNA"] replaceObjectAtIndex:generated withObject:modifiedGene]; // ok!

hmmvot replied:
Загляните в мое решение. 
https://github.com/hmmvot/osx-project-1
В Cell+Mutator.m у меня работает и dna (для доступа к полю объекта), и self.dna, и self->dna.
Может вы делаете что то не так?

А valueForKey убивает весь смысл инкапсуляции (которая в Objective C держится на честном слове разработчика и выполнением им некоторых правил по написанию кода).

slebedev replied:
Посмотрел. И понял, в чём прикол.

Изначально проперти задавался как инструмент, позволяющий на ходу синтезировать не только переменную, с которой можно работать, но и правила доступа извне к этой переменной. Я в своём проекте задавал _только_ проперти. Так как у меня самая последняя версия Xcode, synthesize писать не имело смысла - это делал за меня сам Xcode! И имея следующую строку:

@property (nonatomic, readonly) NSMutableArray *DNA;

Я мог _менять_ значение своей переменной, обращаясь к ней как к _DNA, а извне [myCell DNA] дал бы мне доступ только к геттерам! 

А теперь смотрим на ваш пример: вы использовали внутреннюю переменную NSMutableArray *dna и создали проперти NSArray *dna. И в файле имплементации связали их! @synthesize dna

Я не знаю, одна ли переменная по сути в памяти, или две, но если мне вручную написать @synthesize и использовать так же, как и вы, переменную вместе с проперти, у меня всё получится, да.

hmmvot replied:
Ну все, значит разобрались со взаимным недопониманием.)

hmmvot commented on DNAProject/Cell.h:
```diff
@@ -0,0 +1,19 @@
+@interface Cell : NSObject {
+    NSMutableArray *DNA; //собственно, ДНК
+    NSUInteger capacityOfDNA; //вспомогательная переменная, для хранения количества символов в последовательности
+    NSArray *gene; //набор генов
```
capacityOfDNA и gene хранят одни и те же значения во всех созданных вами клетках. Есть смысл сделать и то и другое константами и не выделять под них память при создании каждой клетки.

slebedev replied:
Каждого созданного объекта, вы хотели сказать? В смысле, клетка=объект-полседовательность ДНК?

hmmvot replied:
Да, именно это.)

slebedev committed:
Небольшие изменения в проекте

Не меняют суть, но делают проект более грамотным. Добавил блок try ...
catch ... finally при вызове метода mutate, генерирую exception в
случае неправильного значения x. Также вынес элементы ячейки
последовательности и количественные параметры, дабы они не создавались
каждый раз при создании нового объекта класса.

slebedev committed:
It works!

Изменения коснулись метода hammingDistance, - как мне резонно заметили,
использовать дополнительные переменные для сравнения элементов - не
слишком уж хорошая затея. Для реализации своей задумки добавил
readonly-свойство, и вручную его синтезировал (чтобы Xcode не
генерировал мне новую переменную, а использовал существующую).
Блок с try ... catch ... finally не стал трогать, хоть мне правильно
сказали, что в таком ключе использовать его не правильно. Это учебный
проект, и в нём нужно показать, как что работает, поэтому я оставил всё
как есть - программа должна рассказывать, где и что не так, и не падать
при удобном случае. Пришлось потратиться на такую "тяжёлую" штуку, как
эксепшн. Заодно почитаю, как использовать NSAssert и иже с ними - будет
полезно.

Не придумал, как можно более удобно использовать свою константную
строку (да, я прекрасно вижу, что я создаю константный указатель, а не
строку, но всё же), и оставил как есть.

slebedev committed:
Предфинальная версия

Как мне справедливо заметили, try ... catch ... finally в данном
примере не показательно. И посоветовали метод NSCAssert, который я
использовал вместо throw.

Алгоритм изменения символов заменил на более оптимальный, на мой
взгляд, такой алгоритм будет делать меньше "лишних" переборов.

И да, не знаю, почему я изначально писал if (![[DNA objectAtIndex:i]
isEqualTo:[[someCell DNA] objectAtIndex:i]]) - я заменил на более
человеческий вид. Смысл от этого, тем не менее, не поменялся.

slebedev committed:
FINAL

"Спрятал" статический массив с набором генов. Обновил комментарии,
добавил в main.m побольше логированного вывода.

YuryRadchenko commented:
При запуске проекта с мутацией величиной 0 выдается ошибка (не просто сообщение!) и программа продолжает работать. Для теста сделал длину ДНК = 5.

Считаю, что мутацию в 0 ячеек может быть, например, когда надо 1 ДНК мутировать, а вторую нет.

2012-11-05 11:48:00.524 DNAProject[4423:403] Формируем первую последовательность ДНК
2012-11-05 11:48:00.528 DNAProject[4423:403] TGCAC
2012-11-05 11:48:00.529 DNAProject[4423:403] Сформировано
2012-11-05 11:48:00.530 DNAProject[4423:403] Формируем вторую последовательность ДНК
2012-11-05 11:48:00.531 DNAProject[4423:403] ACGGT
2012-11-05 11:48:00.532 DNAProject[4423:403] Сформировано
2012-11-05 11:48:00.533 DNAProject[4423:403] Hamming distance: 5
2012-11-05 11:48:00.533 DNAProject[4423:403] Модифицируем первую последовательность
2012-11-05 11:48:00.534 DNAProject[4423:403] **\* Assertion failure in -[Cell(Mutator) mutate:](), /Users/yury/check/osx-project-1.lebedev/DNAProject/main.m:18
2012-11-05 11:48:00.536 DNAProject[4423:403] **\* Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid x value for mutating!'
**\* First throw call stack:
(
    0   CoreFoundation                      0x00007fff83ee6f56 __exceptionPreprocess + 198
    1   libobjc.A.dylib                     0x00007fff85e7bd5e objc_exception_throw + 43
    2   CoreFoundation                      0x00007fff83ee6d8a +[NSException raise:format:arguments:] + 106
    3   Foundation                          0x00007fff8815166c -[NSAssertionHandler handleFailureInFunction:file:lineNumber:description:] + 149
    4   DNAProject                          0x0000000100000fd6 -[Cell(Mutator) mutate:] + 246
    5   DNAProject                          0x000000010000144e main + 350
    6   DNAProject                          0x0000000100000ed4 start + 52
)
terminate called throwing an exception(lldb)

slebedev commented:
В этом случае метод будет вызван только для той последовательности, которая нуждается в мутации, вы так не думаете? Но вообще, можно сделать в проверке вместо строгого равенства нестрогое - х>=0. 0 в данном случае - исключение, которое я решил обработать именно таким образом.

hmmvot commented:
0% - не исключение. Ваш метод принимает проценты, а это число от 0 до 100 (включительно).

YuryRadchenko commented:
Введение уровня мутации 0 не должно давать сообщение, что ошибка. 0 - ничего. Представим, что пользователь вводит значение поочереди для каждой мутации. 0 - значит нет.

Но тут вопрос в другом - почему при 0 программа выводит ошибку (не сообщение) и висит?
Может потому что у меня OS X 10.7 (как я понял, код изначально делался под 10.8) ? XCode у меня 4.5.1

slebedev commented:
Потому что запускаете в Xcode, а он эти вещи отлавливает, работает как бы с процессом в целом. NSCAssert "выталкивает" ошибку, как если бы это была ошибка, например, с памятью.

YuryRadchenko commented:
А в чем надо запускать? Я не сдалал ничего "критического", почему продолжает программа висеть? Недружелюбное поведение программы.

slebedev commented:
Можно запускать в системе, без Xcode - для этого создается исполняемый файл. Это не поведение программы, а поведение Xcode, при запуске программы в этом IDE в случае ошибки запускается debugger, чтобы можно было проанализировать, что к этому привело. Вы когда-нибудь до этого программировали в IDE? В Delphi, например? Там если у запущенного скомпилированного приложения возникнет ошибка, будет то же самое. Нажмите на клавишу "стоп".

YuryRadchenko commented:
> программировали в IDE? В Delphi, 
> Нет.

Но я до сих пор не понимаю, почему на неправильно введенное значение программа пишет отладку и сама не останавливается? 
Если ввел значение правильно:
1. вывод данных
2. программа остановилась. 

Ввел неверно:
1. вывод корректных данных
2. вывод сообщение "неверные данные"
3. вывод множества других строк отладки, которые не дают сразу увидеть предыдущее сообщение.  
4. программа работает дальше и я ее вручную должен останавливать.

Недружелюбно.

YuryRadchenko commented:
Например, в другой проверяемой работе при вводе неверного значения - сообщение что данные ошибочны, расчет корректных данных и остановка программы. Совсем иначе воспринимается.

slebedev commented:
Разные методы используются. Мне преподаватель сказал реализовать так, как вы рассказываете я тоже делал, в одном из коммитов есть.

slebedev commented:
Сразу видно, что вы новичок в программировании. :)

YuryRadchenko commented:
В программировании я больше новичок, чем в юзабилити. Недружественно себя ведет программа, недружественно. 8 баллов из 10 :)

hmmvot commented:
Поясните, как программа без UI может вести себя не дружественно?
С программой вы взаимодействуете сейчас только как программист и то поведение, которое вы получаете при передаче некорректных значений в метод mutate - это очень "дружественное" поведение в отношении к программисту: программа остановилась, зафиксировала состояние и предоставила вам возможность разобраться в чем ошибка.

Так что ваши претензии по этому поводу, мягко говоря, не обоснованны.

slebedev commented:
Вообще-то, с критериями оценки можете ознакомиться на сайте. В 8 баллов не укладывается. Минимум - 9, если счесть тот 0, что я не принимаю в своем решении за корректное значение.

YuryRadchenko commented:
В каком лице проверять? Как чистый программист или как немного пользователь? Со стороны программиста - 10 баллов. Но как пользователь - мне непонятно почему программа "подвисает" при неправильно введенном значении. Учитывая, что программы в итоге пишутся для людей, а не других программистов, я придерживаюсь стороны пользователя.

По критериям оценки. Да, моего критерия вообще нет в наборе, но здравый смысл мне подсказывает, что я могу ввести и свою оценку. Запретов нет на это? Тем более критерии выложили не так давно и мы сейчас все в наработке критериев. 
У меня вопрос в поведении программы при обработке 0% мутации. 0 - отсутствие мутации, а не зависание программы.

8 баллов я выбрал исходя из того что - новый критерий дружелюбности так же имеет 2 балла как и остальные. За него - 0.
Выходит 10 баллов из 12 (это 83%).
Поэтому 8 баллов из 10.

Вобще, тут вопрос в оценке можно долго обсуждать. Нужно, чтобы кто-то из старших вмешался.

hmmvot commented:
Тут не в оценке дело, а в том, что вы эту программу проверяете как "пользователь". Что крайне странно, т.к. пользовательского интерфейса в этой программе нет. И вообще к такой программе стоит относиться как к библиотеке.

Предлагаю вам поставить компании Apple 0 за дружелюбность библиотеки Foundation:

``` objective-c
NSArray *arr = [NSArray arrayWithObjects: @"a", @"b", @"c"];
NSLog(@"%@", [arr objectAtIndex: 10]);
```

Падает с исключением, а не пишет в консоль приятные вещи и корректно завершается.

YuryRadchenko commented:
Нет графического интерфейса. Но интерфейс есть.
Apple не участвует в этом обучении. И предлагаю быть лучше чем Apple :)

Нужно чтобы Рахим вмешался например. А то у нас как в том анекдоте:
Приходят к раввину два человека и просят их рассудить.
Рассказал первый. Раввин говорит:
- Ты прав!
  Второй возмутился:
- Вы еще меня не слышали! Вот слушайте...
  Выслушав второго раввин сказал:
- Ты прав!
  Тут раввина в сторону отводит помощник и говорит:
- Не может быть, чтобы был прав и один и другой.
- Ты прав!

slebedev commented:
То есть преподавателя aspcartman, просматривавшего мой вариант, и поспросившего сделать именно так, вам не хватает?

slebedev commented:
И еще раз, надеюсь, вы поймёте хоть не с первого раза: программа не "подвисает", а при ошибке вмешивается инструмент разработчика - дебаггер, для отладки и выявления причины ошибки. Если программу завершить, то не будет доступа к той памяти, что она использует, не будет доступа к тем переменным, что использовались в программе на момент выявления ошибки. Поэтому программу среда разработки как бы "ставит на паузу".  

Если запустить программу вне среды разработки - она будет работать для пользователя. И завершится с ошибкой Invalid x value. Ничего не зависнет, как вы выразились.

slebedev commented:
И вот еще - эту программу с точки зрения пользователя вообще нельзя оценивать, по одной простой причине: пользователь не может вносить изменения в ход программы. Этот проект вообще можно оценить только как понимание пройденного материала, что выложен в "Свободном университете".  

Нет ничего страшного в том, что вы увидели кучу непонятных строк, ваша задача на конец курса как программиста уметь читать эти строки и исправлять ошибки. Программа сама вам вряд ли сможет сказать больше, чем скажет отладчик.

YuryRadchenko commented:
aspcartman не высказывался по данной конкретной ситуации. Хотелось бы получить его комментарии тогда, чтобы в будущем у нас не было подобных разночтений.

hmmvot commented:
Вам бы поверить на слово и пойти убирать у себя NSLog отовсюду, кроме main, пока не пришел aspcartman и не на ругал за такой подход.)

aspcartman commented:
Равим пришел, встречайте. Единственное, за что я буду карать это за 

> Сразу видно, что вы новичок в программировании. :)

и очень больно.  :)

Оценивать эту программу стоит с точки зрения программиста, а не пользователя.
Наличие NSLog() дело не меняет - есть они или нет, вообще никакой разницы в этой работе.

В реалии ООП всегда должны выполнятся правила инкапсуляции. Чтобы воспользоваться вашим модулем\либой\фреймворком не нужно досканально знать внутреннее устройство модуля\либы\фреймворка. У него должен быть понятный внешний интерфейс, но это не значит, что стоит ставить целью сделать хомячков счастливыми за счет гибкости. 

А еще это значит, что нужно правильно обрабатывать ошибки, и в примере @hmmvot наоборот прекрасно, что приложение рушится при запросе к несуществующему индексу массива - ведь такая ошибка может привести к порче данных, а это вообще ни в какие ворота.

Если где-то дернули mutate с 0 - это норма, ничего не меняем, продолжаем как ни в чем не бывало.
Дернули mutate с -2 или 101 - падаем как можно быстрее. Притом падаем с NSException (сами не обрабатываем), что даст программисту, который будет встраивать наш проект в свой, выбор - падать ли, али обрабатывать. 

Кстати, то, что падает в лог вместе с NSException - крайне полезная информация :)

Однако есть одно большое НО - финальное графическое приложение - не либа, не класс, не фреймворк и к нему уже совершенно другой подход. Вот уж оно не должно падать ни в коем случае, как и не коцать данные. В наш случай - случай модуля, а не приложения.

YuryRadchenko commented:
Всем спасибо. 10 баллов.

slebedev commented:
Спасибо! За что меня карать-то? Я просто увидел, что человек ни разу не работал с IDE и дамп памяти стал страшен. Далее в тексте сказал, что ничего страшного в этом нет, конечный результат будет подразумевать, что всё будет понятно.  

Поясните, пожалуйста.

slebedev committed:
FINAL critical bugfix

Избавился от ненужной константы capacityOfGene, в силу унифицированного
метода класса getRandomGene.

Переписал геттер для своего свойства, иначе возвращалось совсем не то,
что мне нужно было.

В текущей реализации мы имеем вполне себе самостоятельный объект,
который существует по своим собственным законам, повлиять на него можно
только вызвав метод mutate.

slebedev committed:
FINAL critical bugfix

Избавился от ненужной константы capacityOfGene, в силу унифицированного
метода класса getRandomGene.

Переписал геттер для своего свойства, иначе возвращалось совсем не то,
что мне нужно было.

В текущей реализации мы имеем вполне себе самостоятельный объект,
который существует по своим собственным законам, повлиять на него можно
только вызвав метод mutate.

slebedev committed:
FINAL (cosmetic changes - recommended to see)

Итак, по просьбам проверяющих, сделал отдельные файлы для категории.

Модифицировал метод getRandomGene в сторону бОльшей универсальности, он
теперь, помимо генерации случайного символа позволяет заменить символ
на любой другой, отличный от него, что позволило ещё больше
оптимизировать метод mutate.

Добавил возможность вывода сообщений для отладки в методе
getRandomGene: директива debug находится в файле Cell.h, нужны
отладочные сообщения - просто раскомментируйте строчку, иначе не
трогайте - всё итак прекрасно работает.

Решил, что не будет лишним добавить "затычку" для 0% в методе mutate -
понятное дело, что 0% метод смущать не должно, но и остальные объекты
тогда пусть и не создаются.

slebedev committed:
FINAL some changes

slebedev committed:
FINAL testing console git

slebedev committed:
FINAL I can use terminal.app for git, YEAH

slebedev committed:
FINAL algorythm optimisation

slebedev committed:
FINAL release
